File: src/main/java/com/onepointsixtwo/dagger_viewmodel_processor/ViewModelInjectorsClass.java
package com.onepointsixtwo.dagger_viewmodel_processor;


import com.squareup.javapoet.ClassName;
import com.squareup.javapoet.MethodSpec;
import com.squareup.javapoet.TypeName;
import com.squareup.javapoet.TypeSpec;

import java.util.HashMap;
import java.util.Map;

import javax.lang.model.element.Modifier;

public class ViewModelInjectorsClass extends BaseClass {

    private static final String name = "ViewModelInjectors";
    private TypeSpec.Builder typeBuilder;
    private BaseInjectorClass baseInjectorClass;
    private String mapVariableName = "map";
    private TypeName supportActivityType = ClassName.bestGuess("android.support.v4.app.ActivityCompat");
    private TypeName supportFragmentType = ClassName.bestGuess("android.support.v4.app.Fragment");
    private TypeName activityType = ClassName.bestGuess("android.app.Activity");
    private TypeName fragmentType = ClassName.bestGuess("android.app.Fragment");

    public ViewModelInjectorsClass(HashMap<TypeName, TypeName> classToInjectorTypeMapping, BaseInjectorClass baseInjectorClass) {
        this.baseInjectorClass = baseInjectorClass;
        createBasicClass();
        createGetHashmapOfClassToInjectorMethod(classToInjectorTypeMapping);
        createInjectClassActivityMethod();
        createInjectClassFragmentMethod();
        createInjectClassSupportActivityMethod();
        createInjectClassSupportFragmentMethod();
    }

    private void createBasicClass() {
        typeBuilder = TypeSpec.classBuilder("ViewModelInjectors")
                        .addModifiers(Modifier.PUBLIC, Modifier.FINAL);
    }

    /**
     * Creates a method on this class to build a hashmap of the class an injector can inject, to the injector itself mapping
     */
    private void createGetHashmapOfClassToInjectorMethod(HashMap<TypeName, TypeName> classToInjectorTypeMapping) {
        MethodSpec.Builder getInjectorsMethod = MethodSpec.methodBuilder("getInjectorsForClass")
                .addModifiers(Modifier.PRIVATE, Modifier.STATIC);

        //Create the hashmap
        String mapVariableName = "map";
        getInjectorsMethod.addStatement("$T<$T, $T> $L = new $T<>()", hashmapClassName, classTypeName, baseInjectorClass.classType(), mapVariableName, hashmapClassName);

        //Add the values to it - the class maps to the instance of the injector for it
        for(Map.Entry<TypeName, TypeName> entry: classToInjectorTypeMapping.entrySet()) {
            TypeName fragmentOrActivityType = entry.getKey();
            TypeName injectorType = entry.getValue();
            getInjectorsMethod
                    .addStatement("$L.put($T.class, new $T())", mapVariableName, fragmentOrActivityType, injectorType);
        }

        //Handle return statement to return created hashmap
        getInjectorsMethod.addStatement("return $L", mapVariableName);
        getInjectorsMethod.returns(hashmapClassName);

        //Add the method
        typeBuilder.addMethod(getInjectorsMethod.build());
    }

    private void createInjectClassFragmentMethod() {
        String injecteeParamName = "fragment";
        String factoryParamName = "factory";
        String injectorVariableName = "injector";

        MethodSpec.Builder inject = MethodSpec.methodBuilder("inject");
        inject.addModifiers(Modifier.PUBLIC, Modifier.STATIC);

        //Input parameters of an object type and the factory to build view models
        inject.addParameter(fragmentType, injecteeParamName)
                .addParameter(factoryType, factoryParamName);

        //Gets the injector for the given class, checks it is not null, and if it's OK, uses it to inject the class.
        inject.addStatement("$T<$T, $T> $L = getInjectorsForClass()", hashmapClassName, classTypeName, baseInjectorClass.classType(), mapVariableName);
        inject.addStatement("$T $L = $L.get($L.getClass())", baseInjectorClass.classType(), injectorVariableName, mapVariableName, injecteeParamName);
        inject.addStatement("if ($L != null) { $L.inject($L, $L); }", injectorVariableName, injectorVariableName, injecteeParamName, "factory");

        typeBuilder.addMethod(inject.build());
    }

    private void createInjectClassActivityMethod() {
        String injecteeParamName = "activity";
        String factoryParamName = "factory";
        String injectorVariableName = "injector";

        MethodSpec.Builder inject = MethodSpec.methodBuilder("inject");
        inject.addModifiers(Modifier.PUBLIC, Modifier.STATIC);

        //Input parameters of an object type and the factory to build view models
        inject.addParameter(activityType, injecteeParamName)
                .addParameter(factoryType, factoryParamName);

        //Gets the injector for the given class, checks it is not null, and if it's OK, uses it to inject the class.
        inject.addStatement("$T<$T, $T> $L = getInjectorsForClass()", hashmapClassName, classTypeName, baseInjectorClass.classType(), mapVariableName);
        inject.addStatement("$T $L = $L.get($L.getClass())", baseInjectorClass.classType(), injectorVariableName, mapVariableName, injecteeParamName);
        inject.addStatement("if ($L != null) { $L.inject($L, $L); }", injectorVariableName, injectorVariableName, injecteeParamName, "factory");

        typeBuilder.addMethod(inject.build());
    }

    private void createInjectClassSupportFragmentMethod() {
        String injecteeParamName = "supportFragment";
        String factoryParamName = "factory";
        String injectorVariableName = "injector";

        MethodSpec.Builder inject = MethodSpec.methodBuilder("inject");
        inject.addModifiers(Modifier.PUBLIC, Modifier.STATIC);

        //Input parameters of an object type and the factory to build view models
        inject.addParameter(supportFragmentType, injecteeParamName)
                .addParameter(factoryType, factoryParamName);

        //Gets the injector for the given class, checks it is not null, and if it's OK, uses it to inject the class.
        inject.addStatement("$T<$T, $T> $L = getInjectorsForClass()", hashmapClassName, classTypeName, baseInjectorClass.classType(), mapVariableName);
        inject.addStatement("$T $L = $L.get($L.getClass())", baseInjectorClass.classType(), injectorVariableName, mapVariableName, injecteeParamName);
        inject.addStatement("if ($L != null) { $L.inject($L, $L); }", injectorVariableName, injectorVariableName, injecteeParamName, "factory");

        typeBuilder.addMethod(inject.build());
    }

    private void createInjectClassSupportActivityMethod() {
        String injecteeParamName = "supportActivity";
        String factoryParamName = "factory";
        String injectorVariableName = "injector";

        MethodSpec.Builder inject = MethodSpec.methodBuilder("inject");
        inject.addModifiers(Modifier.PUBLIC, Modifier.STATIC);

        //Input parameters of an object type and the factory to build view models
        inject.addParameter(supportActivityType, injecteeParamName)
                .addParameter(factoryType, factoryParamName);

        //Gets the injector for the given class, checks it is not null, and if it's OK, uses it to inject the class.
        inject.addStatement("$T<$T, $T> $L = getInjectorsForClass()", hashmapClassName, classTypeName, baseInjectorClass.classType(), mapVariableName);
        inject.addStatement("$T $L = $L.get($L.getClass())", baseInjectorClass.classType(), injectorVariableName, mapVariableName, injecteeParamName);
        inject.addStatement("if ($L != null) { $L.inject($L, $L); }", injectorVariableName, injectorVariableName, injecteeParamName, "factory");

        typeBuilder.addMethod(inject.build());
    }


    @Override
    TypeName classType() {
        return ClassName.bestGuess(name);
    }

    @Override
    protected TypeSpec.Builder getTypeBuilder() {
        return typeBuilder;
    }
}

File: src/main/java/com/onepointsixtwo/dagger_viewmodel_processor/ViewModelInjectorsClass.kt
package com.onepointsixtwo.dagger_viewmodel_processor


import com.squareup.kotlinpoet.*
import com.squareup.kotlinpoet.ParameterizedTypeName.Companion.parameterizedBy

import java.util.HashMap

class ViewModelInjectorsClass(
    classToInjectorTypeMapping: HashMap<TypeName, TypeName>,
    private val baseInjectorClass: BaseInjectorClass
) : BaseClass() {

    protected override var typeBuilder: TypeSpec.Builder? = null
        private set
    private val mapVariableName = "map"
    private val supportActivityType = ClassName.bestGuess("androidx.core.app.ActivityCompat")
    private val supportFragmentType = ClassName.bestGuess("androidx.fragment.app.Fragment")
    private val activityType = ClassName.bestGuess("android.app.Activity")
    private val fragmentType = ClassName.bestGuess("android.app.Fragment")

    init {
        createBasicClass()
        createGetHashmapOfClassToInjectorMethod(classToInjectorTypeMapping)
        createInjectClassActivityMethod()
        createInjectClassFragmentMethod()
        createInjectClassSupportActivityMethod()
        createInjectClassSupportFragmentMethod()
    }

    private fun createBasicClass() {
        typeBuilder = TypeSpec.objectBuilder("ViewModelInjectors")
            .addModifiers(KModifier.PUBLIC)
    }

    /**
     * Creates a method on this class to build a hashmap of the class an injector can inject, to the injector itself mapping
     */
    private fun createGetHashmapOfClassToInjectorMethod(classToInjectorTypeMapping: HashMap<TypeName, TypeName>) {
        val getInjectorsMethod = FunSpec.builder("getInjectorsForClass")
            .addModifiers(KModifier.PRIVATE)

        //Create the hashmap
        val mapVariableName = "map"
        getInjectorsMethod.addStatement(
            "val %N = mutableMapOf<%T<*>, %T>()",
            mapVariableName,
            classTypeName,
            baseInjectorClass.classType()
        )

        //Add the values to it - the class maps to the instance of the injector for it
        for ((fragmentOrActivityType, injectorType) in classToInjectorTypeMapping) {
            getInjectorsMethod
                .addStatement(
                    "%N.put(%T::class, %T())",
                    mapVariableName,
                    fragmentOrActivityType,
                    injectorType
                )
        }

        //Handle return statement to return created hashmap
        getInjectorsMethod.addStatement("return %N", mapVariableName)

        val returnType = Map::class.asClassName().parameterizedBy(classTypeName, baseInjectorClass.classType())
        getInjectorsMethod.returns(returnType)

        //Add the method
        typeBuilder!!.addFunction(getInjectorsMethod.build())
    }

    private fun createInjectClassFragmentMethod() {
        val injecteeParamName = "fragment"
        val factoryParamName = "factory"
        val injectorVariableName = "injector"

        val inject = FunSpec.builder("inject")
        inject.addModifiers(KModifier.PUBLIC)

        //Input parameters of an object type and the factory to build view models
        inject.addParameter(injecteeParamName, fragmentType)
            .addParameter(factoryParamName, factoryType)

        //Gets the injector for the given class, checks it is not null, and if it's OK, uses it to inject the class.
        inject.addStatement(
            "\$T<\$T, \$T> \$L = getInjectorsForClass()",
            hashmapClassName,
            classTypeName,
            baseInjectorClass.classType(),
            mapVariableName
        )
        inject.addStatement(
            "\$T \$L = \$L.get(\$L.getClass())",
            baseInjectorClass.classType(),
            injectorVariableName,
            mapVariableName,
            injecteeParamName
        )
        inject.addStatement(
            "if (\$L != null) { \$L.inject(\$L, \$L); }",
            injectorVariableName,
            injectorVariableName,
            injecteeParamName,
            "factory"
        )

        typeBuilder!!.addFunction(inject.build())
    }

    private fun createInjectClassActivityMethod() {
        val injecteeParamName = "activity"
        val factoryParamName = "factory"
        val injectorVariableName = "injector"

        val inject = FunSpec.builder("inject")
        inject.addModifiers(KModifier.PUBLIC)

        //Input parameters of an object type and the factory to build view models
        inject.addParameter(injecteeParamName, activityType)
            .addParameter(factoryParamName, factoryType)

        //Gets the injector for the given class, checks it is not null, and if it's OK, uses it to inject the class.
        inject.addStatement(
            "\$T<\$T, \$T> \$L = getInjectorsForClass()",
            hashmapClassName,
            classTypeName,
            baseInjectorClass.classType(),
            mapVariableName
        )
        inject.addStatement(
            "\$T \$L = \$L.get(\$L.getClass())",
            baseInjectorClass.classType(),
            injectorVariableName,
            mapVariableName,
            injecteeParamName
        )
        inject.addStatement(
            "if (\$L != null) { \$L.inject(\$L, \$L); }",
            injectorVariableName,
            injectorVariableName,
            injecteeParamName,
            "factory"
        )

        typeBuilder!!.addFunction(inject.build())
    }

    private fun createInjectClassSupportFragmentMethod() {
        val injecteeParamName = "supportFragment"
        val factoryParamName = "factory"
        val injectorVariableName = "injector"

        val inject = FunSpec.builder("inject")
        inject.addModifiers(KModifier.PUBLIC, KModifier.COMPANION)

        //Input parameters of an object type and the factory to build view models
        inject.addParameter(injecteeParamName, supportFragmentType)
            .addParameter(factoryParamName, factoryType)

        //Gets the injector for the given class, checks it is not null, and if it's OK, uses it to inject the class.
        inject.addStatement(
            "\$T<\$T, \$T> \$L = getInjectorsForClass()",
            hashmapClassName,
            classTypeName,
            baseInjectorClass.classType(),
            mapVariableName
        )
        inject.addStatement(
            "\$T \$L = \$L.get(\$L.getClass())",
            baseInjectorClass.classType(),
            injectorVariableName,
            mapVariableName,
            injecteeParamName
        )
        inject.addStatement(
            "if (\$L != null) { \$L.inject(\$L, \$L); }",
            injectorVariableName,
            injectorVariableName,
            injecteeParamName,
            "factory"
        )

        typeBuilder!!.addFunction(inject.build())
    }

    private fun createInjectClassSupportActivityMethod() {
        val injecteeParamName = "supportActivity"
        val factoryParamName = "factory"
        val injectorVariableName = "injector"

        val inject = FunSpec.builder("inject")
        inject.addModifiers(KModifier.PUBLIC)

        //Input parameters of an object type and the factory to build view models
        inject.addParameter(injecteeParamName, supportActivityType)
            .addParameter(factoryParamName, factoryType)

        //Gets the injector for the given class, checks it is not null, and if it's OK, uses it to inject the class.
        inject.addStatement(
            "\$T<\$T, \$T> \$L = getInjectorsForClass()",
            hashmapClassName,
            classTypeName,
            baseInjectorClass.classType(),
            mapVariableName
        )
        inject.addStatement(
            "\$T \$L = \$L.get(\$L.getClass())",
            baseInjectorClass.classType(),
            injectorVariableName,
            mapVariableName,
            injecteeParamName
        )
        inject.addStatement(
            "if (\$L != null) { \$L.inject(\$L, \$L); }",
            injectorVariableName,
            injectorVariableName,
            injecteeParamName,
            "factory"
        )

        typeBuilder!!.addFunction(inject.build())
    }


    internal override fun classType(): TypeName {
        return ClassName.bestGuess(name)
    }

    override fun fileName(): String {
        return name
    }

    companion object {

        private val name = "ViewModelInjectors"
    }
}
